pkg/executables/cmk.go

```diff
@@ -0,0 +1,371 @@
+		}
+		affinityGroup := response.CmkAffinityGroups
+		if len(affinityGroup) > 1 {
+			return fmt.Errorf("duplicate affinity group %s found", affinityGroupId)
+		} else if len(affinityGroup) == 0 {
+			return fmt.Errorf("affinity group %s not found", affinityGroupId)
+		}
+	}
+	return nil
+}
+
+func (c *Cmk) ValidateZonePresent(ctx context.Context, zone string) error {
+	result, err := c.execWithNameAndIdFilters(ctx, zone, "list", "zones")
+	if err != nil {
+		return fmt.Errorf("error getting zones info: %v", err)
+	}
+	if result.Len() == 0 {
+		return fmt.Errorf("zone %s not found", zone)
+	}
+
+	response := struct {
+		CmkZones []CmkZone `json:"zone"`
+	}{}
+	err = json.Unmarshal(result.Bytes(), &response)
+	if err != nil {
+		return fmt.Errorf("failed to parse response into json: %v", err)
+	}
+	zones := response.CmkZones
+	if len(zones) > 1 {
+		return fmt.Errorf("duplicate zone %s found", zone)
+	} else if len(zones) == 0 {
+		return fmt.Errorf("zone %s not found", zone)
+	}
+	return nil
+}
+
+// TODO: Add support for domain filtering
+func (c *Cmk) ValidateAccountPresent(ctx context.Context, account string) error {
+	result, err := c.execWithNameAndIdFilters(ctx, account, "list", "accounts")
+	if err != nil {
+		return fmt.Errorf("error getting accounts info: %v", err)
+	}
+	if result.Len() == 0 {
+		return fmt.Errorf("account %s not found", account)
+	}
+
+	response := struct {
+		CmkAccounts []CmkAccount `json:"account"`
+	}{}
+	err = json.Unmarshal(result.Bytes(), &response)
+	if err != nil {
+		return fmt.Errorf("failed to parse response into json: %v", err)
+	}
+	accounts := response.CmkAccounts
+	if len(accounts) > 1 {
+		return fmt.Errorf("duplicate account %s found", account)
+	} else if len(accounts) == 0 {
+		return fmt.Errorf("account %s not found", account)
+	}
+	return nil
+}
+
+// cmkExecConfig contains transient information for the execution of cmk commands
+// It must be cleaned after each execution to prevent side effects from past executions options
+type cmkExecConfig struct {
+	env        map[string]string
+	ConfigFile string
+}
+
+func NewCmk(executable Executable, writer filewriter.FileWriter) *Cmk {
+	return &Cmk{
+		writer:     writer,
+		executable: executable,
+		retrier:    retrier.NewWithMaxRetries(maxRetriesCmk, backOffPeriodCmk),
+	}
+}
+
+// ValidateCloudStackConnection Calls `cmk sync` to ensure that the endpoint and credentials + domain are valid
+func (c *Cmk) ValidateCloudStackConnection(ctx context.Context) error {
+	buffer, err := c.exec(ctx, "sync")
+	if err != nil {
+		return fmt.Errorf("error validating cloudstack connection for cmk config %s: %v", buffer.String(), err)
+	}
+	logger.MarkPass("Connected to CloudStack server")
+	return nil
+}
+
+func (c *Cmk) execWithNameAndIdFilters(ctx context.Context, parameterValue string, genericArgs ...string) (stdout bytes.Buffer, err error) {
+	argsWithNameFilterArg := append(genericArgs, fmt.Sprintf("name=\"%s\"", parameterValue))
```

Comment:
I am a little confused by the name `parameterValue`. Can we just call it `name` or `resourceName`? Also, I think this function is better named `execWithNameWithIdFiltersRetry` to describe what it is doing, but maybe that is confusing too.

Reply:
The reasoning for this naming is we're unsure at this point if the input value is a resource name or unique ID. I thought "parameter" might be the right word to convey that ambiguity, so I'm inclined to push back against a variable name including `name`. I'm happy to rename the function, however - I think that's a good idea. Let me know if you disagree

Reply:
@maxdrib do all the `Validate...` methods take a name or ID? Or some of them only take names and others only take IDs?

Reply:
It varies based on what CloudMonkey supports and what usecases we support. For example with affinity groups, CAPC only supports passing in a list of UUID's, not names, so we only query cmk by ID. I think the opposite is true if/when we validate the network (which is a TODO). There's [no parameter for "name" when listing networks](https://cloudstack.apache.org/api/apidocs-4.14/apis/listNetworks.html), so we can only query by ID

Reply:
Got it. In that case I would suggest having separate methods for ID and name. If there is a type of resource that needs both, I think that's totally ok.